### PR TITLE
Implement DatasetAuthorization dependency, add to endpoints

### PR DESCRIPTION
**Pull request by tcnichol**
This pull request adds dataset authorization dependency and adds this to the dataset endpoints.

### CODE REVIEW

File: backend/app/deps/authorization_deps.py [hunk 1]
```diff
@@ -92,6 +92,41 @@ async def __call__(
             )
 
 
+class DatasetAuthorization:
```
Comment (max-zilla) [inline backend/app/deps/authorization_deps.py:95]: The base Authorization class that Luigi implemented functions as DatasetAuthorization, you should be able to remove this one and use Authorization instead.

File: backend/app/routers/datasets.py [hunk 2]
```diff
@@ -358,6 +360,8 @@ async def edit_dataset(
     db: MongoClient = Depends(dependencies.get_db),
     user_id=Depends(get_user),
     es=Depends(dependencies.get_elasticsearchclient),
+    allow: bool = Depends(DatasetAuthorization("editor")),
```
Comment (max-zilla) [inline backend/app/routers/datasets.py:363]: in these cases, you can just use Depends(Authorization( as mentioned above.

File: backend/app/routers/datasets.py [hunk 3]
```diff
@@ -498,6 +506,7 @@ async def add_folder(
     folder_in: FolderIn,
     user=Depends(get_current_user),
     db: MongoClient = Depends(dependencies.get_db),
+    allow: bool = Depends(DatasetAuthorization("editor")),
```
Comment (max-zilla) [inline backend/app/routers/datasets.py:509]: there is also a role between viewer and editor that is "uploader". I think a few of these could use the more relaxed uploader role instead if they involve only additive things - maybe add_folder, save_file, get_dataset_extract?

File: backend/app/routers/datasets.py [hunk 4]
```diff
@@ -557,6 +567,7 @@ async def delete_folder(
     db: MongoClient = Depends(dependencies.get_db),
     fs: Minio = Depends(dependencies.get_fs),
     es: Elasticsearch = Depends(dependencies.get_elasticsearchclient),
+    allow: bool = Depends(DatasetAuthorization("owner")),
```
Comment (max-zilla) [inline backend/app/routers/datasets.py:570]: i think it is OK to let editors do this also

**Review by max-zilla — CHANGES_REQUESTED** [review]

File: backend/app/routers/datasets.py (hunk 4 above)
Comment (tcnichol) [reply backend/app/routers/datasets.py]: These are fixed. I don't use the allow yet but I am working on that, will have some questions at the meeting tomorrow.

File: backend/app/routers/datasets.py (hunk 2 above)
Comment (tcnichol) [reply backend/app/routers/datasets.py]: Fixed this.

File: backend/app/deps/authorization_deps.py (hunk 1 above)
Comment (tcnichol) [reply backend/app/deps/authorization_deps.py]: Removed.

File: backend/app/routers/datasets.py (hunk 3 above)
Comment (tcnichol) [reply backend/app/routers/datasets.py]: I put uploader for these.

File: backend/app/routers/datasets.py [hunk 5]
```diff
@@ -358,7 +360,10 @@ async def edit_dataset(
     db: MongoClient = Depends(dependencies.get_db),
     user_id=Depends(get_user),
     es=Depends(dependencies.get_elasticsearchclient),
+    allow: bool = Depends(Authorization("editor")),
 ):
+    if not allow:
+        raise HTTPException(status_code=404, detail=f"Dataset {dataset_id} not found")
```
Comment (longshuicy) [inline backend/app/routers/datasets.py:366]: 404 here intended? not 403?

Comment (tcnichol) [reply backend/app/routers/datasets.py:366]: Which should these be? I can change them all to 403.

**Review by max-zilla — APPROVED** [review]

**Pull request merged**